File: tests/fixtures/safe/secure-stretch.go
package main

import (
	"crypto/sha256"
	"crypto/tls"
	"log"
	"net/http"
	"os"
)

// Safe: credentials come from the environment
func connect() string {
	password := os.Getenv("DB_PASSWORD")
	return password
}

// Safe: static path
func readConfig() ([]byte, error) {
	return os.ReadFile("/etc/app/config.yml")
}

// Safe: static URL
func healthCheck() (*http.Response, error) {
	return http.Get("https://status.example.com/health")
}

// Safe: strong hash algorithm
func hashPassword(password string) [32]byte {
	return sha256.Sum256([]byte(password))
}

// Safe: non-sensitive log message
func logStartup() {
	log.Println("server started")
}

// Safe: TLS verification enabled
func secureClient() *tls.Config {
	return &tls.Config{InsecureSkipVerify: false}
}

File: tests/fixtures/safe/SecureStretch.java
package demo;

import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class SecureStretch {

    // Safe: credentials come from the environment
    public String connect() {
        String password = System.getenv("DB_PASSWORD");
        return password;
    }

    // Safe: static path
    public byte[] readConfig() throws Exception {
        return Files.readAllBytes(Paths.get("/etc/app/config.yml"));
    }

    // Safe: user path is normalized and prefix-checked before use
    public byte[] readUpload(String name) throws Exception {
        Path target = Paths.get("/data/uploads/" + name).normalize();
        if (!target.startsWith("/data/uploads")) throw new SecurityException("bad path");
        return Files.readAllBytes(target);
    }

    // Safe: static URL
    public URL healthCheck() throws Exception {
        return new URL("https://status.example.com/health");
    }
}
